contracts/MarketOracle.sol

```diff
@@ -14,104 +14,128 @@ interface IOracle {
 /**
  * @title Market Oracle
  *
- * @dev Provides the aggregated exchange rate data onchain using data from a whitelisted
- *      set of market source contracts.
- *      Exchange rate is the TOKEN:TARGET rate.
+ * @dev Provides the aggregated rate data onchain using data from a whitelisted
+ *      set of providers.
  */
 contract MarketOracle is Ownable, IOracle {
     using SafeMath for uint256;
 
-    // Whitelist of sources
-    MarketSource[] public whitelist;
+    struct Report {
+        address provider;
+        uint256 timestamp;
+        uint256 payload;
+    }
 
-    event LogSourceAdded(MarketSource source);
-    event LogSourceRemoved(MarketSource source);
-    event LogSourceExpired(MarketSource source);
+    mapping (address => uint256) internal whitelistMap;
```

Comment:
AFAICT, the value represents 1+index in the reports array, so we can differentiate addresses that don't exist in the map on lookup.

What if instead our storage was just `mapping (address => Report)` ?
Any invalid address would be a zeroed out report and be expired.
Or we could add a "valid" bool in the report to make it explicit.

That would simplify the logic a lot and maybe save storage.

Reply:
Yes that is the optimal way but  there is no way to loop on keys of a hashmap in solidity. Another alternative is to store mapping (address => Report) and auxiliary list of addresses. But then removing addresses require loop on list of address.

Reply:
I guess the last approach would be to have a flat array of reports, and loop through while checking the provider.

I don't know which would be more efficient. Your way would most likely get relatively more efficient as number of providers increases. If the flat array is more efficient with 1 provider, I wonder where the crossing point is? 🤷‍♂️

Reply:
Costwise for PushReport function. 
A) Current version: Read + 2 Writes
B) The alternative in my comment: 2 Writes
C) The alternative in your comment: P reads + 2 Writes

The crossover point betn C and (A/B) is  2 providers. and for 1 provider it will be more or less the same.
Costwise B is a clear winner.

Reply:
I am changing the code to B)